### PR TITLE
feat: Firestore default database support

### PR DESCRIPTION
**Pull request by rwhogg**
This includes support for both Firestore Native and Datastore mode databases.

Part of https://github.com/hashicorp/terraform-provider-google/issues/12294

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_firestore_database
```

### CODE REVIEW

**Comment by rwhogg** [comment]
The basics are there, but at the moment it's kind of random whether the API calls (specifically the GET to the created database) succeed or fail with a 403.

Any tips for resolving this issue?

File: mmv1/products/firestore/api.yaml
Comment (pl04351820) [inline mmv1/products/firestore/api.yaml:44]: q: Will Terraform wait for LRO to finish? 

The existing API behavior is 
- CreateDatabase will return done = false in initial resp. 
- The client can understand the LRO status (done = true) by fetching with operation_id.

File: mmv1/products/firestore/api.yaml
Comment (pl04351820) [inline mmv1/products/firestore/api.yaml:108]: Consider exposing create_time (OUTPUT_ONLY)?

File: mmv1/products/firestore/terraform.yaml
Comment (pl04351820) [inline mmv1/products/firestore/terraform.yaml:26]: Consider removing `default`. In Terraform layer, database_id is transparent so `(default)` vs `test-db` do not make any difference.

File: mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb
Comment (pl04351820) [inline mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb:13]: Consider adding a test for create surface.

File: mmv1/products/firestore/api.yaml
Comment (pl04351820) [inline mmv1/products/firestore/api.yaml:30]: q: Do we need another implementation for support `import Terraform configuration from the existing GCP project`? Or this is auto generated after we implement the surface

File: mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb
Comment (rwhogg) [reply mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb:13]: The examples autogenerate tests that handle database creation. It's only necessary to handwrite the update tests.

File: mmv1/products/firestore/terraform.yaml
Comment (rwhogg) [reply mmv1/products/firestore/terraform.yaml:26]: I think I'll leave this in for now. I'd like to emphasize that this example is about creating the (default) database.

Comment (rwhogg) [reply mmv1/products/firestore/terraform.yaml:26]: On second thought, you're right, there's nothing too special about this. I'll change the name.

File: mmv1/products/firestore/api.yaml
Comment (rwhogg) [reply mmv1/products/firestore/api.yaml:44]: Yes, it will wait until the LRO's are done.

If you run the tests with TF_LOG=trace, you can see it polling the LRO for completion.

File: mmv1/products/firestore/api.yaml
Comment (rwhogg) [reply mmv1/products/firestore/api.yaml:30]: Well, importing does actually require a bit of handwriting in this case, but I have already implemented it. So it should already work.

**Comment by rwhogg** [comment]
@roaks3 I think everything should be passing now

**Comment by rwhogg** [comment]
FYI @roaks3 the failing test there is a known issue: https://github.com/hashicorp/terraform-provider-google/issues/13430

File: mmv1/products/firestore/api.yaml
Comment (rainshen49) [inline mmv1/products/firestore/api.yaml:34]: Should this use `database_id` instead? The name is then a separate output_only property that gives the full resource name if we want to be consistent with https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases#resource:-database

**Comment by roaks3** [comment]
@rwhogg sorry for the delay, I did see your note and am planning to look at this first thing tomorrow morning

File: mmv1/products/firestore/api.yaml
Comment (roaks3) [inline mmv1/products/firestore/api.yaml:55]: Super nit: I'm seeing different casing for the true/false values in these yaml files. I suspect some of this was copy-and-pasted, and it looks like many of our existing files already have this mixed casing, so up to you if you want to make them consistent.

File: mmv1/products/firestore/api.yaml
Comment (roaks3) [inline mmv1/products/firestore/api.yaml:106]: Do we need Terraform to set the default to disabled, or could we possibly rely on the server to use a default? Reducing defaults in Terraform generally makes things easier for us to maintain and change if necessary.

File: mmv1/products/firestore/api.yaml
Comment (roaks3) [inline mmv1/products/firestore/api.yaml:113]: Deletion protection is a whole topic for our team at the moment. I'll go back and make sure I'm caught up on current recommendations, but here I believe we generally want the default to be enabled, to protect unknowing users. Do you know if you need to keep this disabled by default?

File: mmv1/templates/terraform/decoders/firestore_database.go.erb
Comment (roaks3) [inline mmv1/templates/terraform/decoders/firestore_database.go.erb:1]: This looks like it could potentially be replaced with a generic "flatten" function, but I'm not quite sure which. Perhaps `name_from_self_link` is enough? Do you know if there is special handling you need for this resource?

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (roaks3) [inline mmv1/templates/terraform/examples/firestore_database.tf.erb:2]: Nit: There is some inconsistent indentation in these example files, which will show up in the resource documentation website

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (roaks3) [inline mmv1/templates/terraform/examples/firestore_database.tf.erb:29]: We may want to use a variable here and pass it in from terraform.yaml, so that the example can be `(default)`, but the test can be something that we can identify and clean up, like `tf-test-database`. Otherwise, we might run into issues with tests where this database already exists (if this is a singleton resources, then perhaps this doesn't apply).

**Review by roaks3 — COMMENTED** [review]
Looks good overall! I left some comments that are not necessarily blocking, but I'll wait to hear back on your responses.

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (rwhogg) [reply mmv1/templates/terraform/examples/firestore_database.tf.erb:29]: As of this writing, "(default)" is the only supported database ID.

File: mmv1/products/firestore/api.yaml
Comment (rwhogg) [reply mmv1/products/firestore/api.yaml:55]: I'm happy to make them consistent. I like all-lowercase for boolean constants, so unless you're aware of a preference for `True` I'll switch all of my new code to use all-lowercase.

File: mmv1/products/firestore/api.yaml
Comment (rwhogg) [reply mmv1/products/firestore/api.yaml:106]: I suspect the default can be safely removed here. I will try. Thanks for the suggestion!

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (rwhogg) [reply mmv1/templates/terraform/examples/firestore_database.tf.erb:29]: Also worth noting that database deletion is not currently supported, which is why all the tests spin up their own Cloud project, enable the Firestore API, create/update the database, and then delete the whole thing.

File: mmv1/products/firestore/api.yaml
Comment (rwhogg) [reply mmv1/products/firestore/api.yaml:113]: I'll take delete protection out of the PR for now.

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (rwhogg) [reply mmv1/templates/terraform/examples/firestore_database.tf.erb:34]: For now, I will remove delete protection from this PR. We can revisit that later.

File: mmv1/templates/terraform/encoders/firestore_database.go.erb
Comment (rwhogg) [reply mmv1/templates/terraform/encoders/firestore_database.go.erb]: Yup, that works. Thanks!

File: mmv1/templates/terraform/decoders/firestore_database.go.erb
Comment (rwhogg) [reply mmv1/templates/terraform/decoders/firestore_database.go.erb]: Yup, that works. Thanks!

**Comment by rwhogg** [comment]
Repushed, thanks a bunch for the advice!

File: mmv1/templates/terraform/examples/firestore_database.tf.erb
Comment (roaks3) [reply mmv1/templates/terraform/examples/firestore_database.tf.erb]: That works, thanks for explaining!

File: mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb:23]: I might not understand what is happening here, but do you know if you need this line? I didn't see the `random` provider being used by your test.

**Review by roaks3 — APPROVED** [review]
One final comment on the `random` provider, but approving because the rest of this looks good to me.

File: mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb
Comment (rwhogg) [reply mmv1/third_party/terraform/tests/resource_firestore_database_update_test.go.erb]: Not needed after all, thanks!

**Comment by rwhogg** [comment]
Pushed and all matters resolved; unfortunately some unrelated tests failed.

**Review by roaks3 — APPROVED** [review]

**Comment by roaks3** [comment]
Yea, those tests are unrelated. @rwhogg if you aren't waiting for reviews from any other team members I can go ahead and merge.

**Comment by rwhogg** [comment]
I'm not waiting for anyone else. So yes, please merge. Thanks a bunch :)

**Pull request merged**
